### PR TITLE
build: cross-margin partial margin (pt1)

### PR DESCRIPTION
**Pull request by dsshap**
## Summary

Implemented first phase of partial margining.
Changing whitelist method from `engineAccess` to `isAllowed`

## Details

Set masks for assets that are cross margin-able. 
On `_isAccountAboveWater` check against deposited collateral. 
Removing Mint constraint that all PUTs need to be collateralized in strike asset and CALLs with underlying asset.

## Checklist

Ensure you completed **all of the steps** below before submitting your pull request:

- [x] Add [natspec](https://docs.soliditylang.org/en/latest/natspec-format.html) for all functions / parameters
- [x] Ran `forge snapshot`
- [x] Ran `forge fmt`
- [x] Ran `forge test`
- [ ] [Triage Slither issues](../README.md#triage-issues), and post uncertain ones in the PR

### CODE REVIEW

File: src/test/engine-integrations/cross-margin/MintTokenWithPartialMarginBeta.sol [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+import "../../../config/errors.sol";
+
+import "../../../core/engines/cross-margin/types.sol";
+
+import "../../../test/mocks/MockERC20.sol";
+
+// solhint-disable-next-line contract-name-camelcase
+contract TestMintWithPartialMarginBeta_CM is CrossMarginFixture {
+    MockERC20 internal lsEth;
+    MockERC20 internal mmc;
```
Comment (antoncoding) [inline src/test/engine-integrations/cross-margin/MintTokenWithPartialMarginBeta.sol:19]: what does lsEth and mmc stands for?

File: src/test/engine-integrations/cross-margin/MintTokenWithPartialMarginBeta.sol [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        oracle.setSpotPrice(address(weth), 3000 * UNIT);
+    }
+
+    function testRemovePartialMarginMask() public {
+        assertEq(engine.getPartialMarginMask(address(weth), address(lsEth)), true);
+
+        engine.setPartialMarginMask(address(weth), address(lsEth), false);
+
+        assertEq(engine.getPartialMarginMask(address(weth), address(lsEth)), false);
+    }
```
Comment (antoncoding) [inline src/test/engine-integrations/cross-margin/MintTokenWithPartialMarginBeta.sol:78]: can you also add tests that setting to false doesn't affect other pairs?

File: src/core/engines/cross-margin/CrossMarginEngine.sol [hunk 3]
```diff
@@ -196,6 +211,24 @@ contract CrossMarginEngine is
...
+     * @notice  sets the Partial Margin Mask for a pair of assets
+     * @param _assetX the id of the asset a
+     * @param _assetY the id of the asset b
+     * @param _value is margin-able
+     */
+    function setPartialMarginMask(address _assetX, address _assetY, bool _value) external {
+        _checkOwner();
+
+        uint256 collateralId = grappa.assetIds(_assetX);
+        uint256 mask = 1 << (grappa.assetIds(_assetY) & 0xff);
```
Comment (antoncoding) [inline src/core/engines/cross-margin/CrossMarginEngine.sol:224]: what is the purpose of `&0xff`? isn't it just gonna be the same value?

File: src/core/engines/cross-margin/CrossMarginEngine.sol [hunk 4]
```diff
@@ -292,17 +334,48 @@ contract CrossMarginEngine is
...
 
-            unchecked {
-                ++i;
+                for (y; y < collaterals.length; ++y) {
+                    // only setting amount on first pass, dont need to repeat each inner loop
+                    if (x == 0) amounts[y] = collaterals[y].amount;
+
+                    uint8 collateralId = collaterals[y].collateralId;
+
+                    // setting mask to 1 if collateralId is
```
Comment (antoncoding) [inline src/core/engines/cross-margin/CrossMarginEngine.sol:356]: this comment seems unfinished 🤣

**Review by antoncoding — APPROVED** [review]
Some minor comments

File: src/test/engine-integrations/cross-margin/MintTokenWithPartialMarginBeta.sol (hunk 1 above)
Comment (dsshap) [reply src/test/engine-integrations/cross-margin/MintTokenWithPartialMarginBeta.sol:19]: LsETH - liquid staked eth
mmc - money market type coin

File: src/core/engines/cross-margin/CrossMarginEngine.sol (hunk 3 above)
Comment (dsshap) [reply src/core/engines/cross-margin/CrossMarginEngine.sol:224]: Taken from the oz bitmap implementation

File: src/test/engine-integrations/cross-margin/MintTokenWithPartialMarginBeta.sol (hunk 1 above)
Comment (antoncoding) [reply src/test/engine-integrations/cross-margin/MintTokenWithPartialMarginBeta.sol:19]: better add them somewhere if you haven't

Comment (dsshap) [reply src/test/engine-integrations/cross-margin/MintTokenWithPartialMarginBeta.sol:19]: What do you mean?

Comment (antoncoding) [reply src/test/engine-integrations/cross-margin/MintTokenWithPartialMarginBeta.sol:19]: just meant comment them somewhere in the code

**Pull request merged**